runtime/j9vm/valhallavmi.cpp

```diff
@@ -91,24 +110,33 @@ JVM_NewNullRestrictedArray(JNIEnv *env, jclass componentType, jint length)
 		goto done;
 	}
 
-	if (!(J9_IS_J9CLASS_VALUETYPE(ramClass) && J9_IS_J9CLASS_ALLOW_DEFAULT_VALUE(ramClass))) {
+	if (!J9_IS_J9CLASS_VALUETYPE(ramClass)) {
 		vmFuncs->setCurrentException(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, NULL);
 		goto done;
 	}
 
-	if (NULL == J9CLASS_GET_NULLRESTRICTED_ARRAY(ramClass)) {
+	if (isNullRestricted && !J9_IS_J9CLASS_ALLOW_DEFAULT_VALUE(ramClass)) {
+		vmFuncs->setCurrentException(currentThread, J9VMCONSTANTPOOL_JAVALANGILLEGALARGUMENTEXCEPTION, NULL);
+		goto done;
+	}
+
```

Comment:
Maybe something like the following up front, then you dont need the ternary op in the block

```
arrayClass = ramClass->arrayClass;
options = 0;

if (isNullRestricted) {
    arrayClass = J9CLASS_GET_NULLRESTRICTED_ARRAY(ramClass);
    options = J9_FINDCLASS_FLAG_CLASS_OPTION_NULL_RESTRICTED_ARRAY;
}

...

and then 

`arrayClass = vmFuncs->internalCreateArrayClassWithOptions` 


```